### DISCUSSION TITLE
Allow loaders to be dynamically loaded

### DISCUSSION THREAD

**Discussion by renchap** [libvips/libvips#2018] [Ideas]
The initial discussion is here: https://github.com/libvips/ruby-vips/issues/176

The various loaders have different levels of maturity, testing and dependencies. Currently the loaders are configured during the build and built into the monolithic `libvips` library. As an example, currently the `libvips` Debian packages pulls 700 MB of dependencies, including Firefox.

It would be very useful to have a way to distribute the loaders in separate packages, so distributions can provide only "core" loaders (the most tested, safe and used ones) in the  `libvips` package, and provide additional packages for less-common loaders. The benefits for this are:
- security: currently most distro-provided packages are using many loaders, and a security issue in any of those (or their deps) will affect vips users
- stability
- fewer dependencies: if I only want to process "web" formats, I do not expect Firefox to be pulled, or 700MB or packages to be installed
- extensibility: you should be able to build your own loader and load it at runtime, useful if you need a custom one or during loader development

**Comment by jcupitt** [comment]
I think the huge list of dependancies on Debian is an anomaly: the Fedora package is much slimmer, for example. The latest version of the Debian package has been revised and is much smaller now.

However I agree, this would make building and configuring libvips simpler and safer.

libvips already has a simple plugin system based on GModule, so it would be easy to move the uncommon loaders to a separate binary, built in a separate project, with no changes to user code.

I don't think this would work on Windows. My understanding is that DLLs do not support backlinking, so a plugin is not able to make direct calls to functions in the library it is loaded from. Instead, the caller should pass in a big struct holding pointers to every function in the API.

The simplest solution would probably be to build the extra modules directly into the libvips binary on Windows. This would need some changes to the various Windows build systems.

As a first step, let's make a libvips branch with the less-well-tested loaders stripped out, and try making a separate plugin for them.

**Comment by jcupitt** [comment]
Your extensibility bulletpoint is possible now, you just build it as a plugin and put it in the vips lib folder.

Here's openslideload as a plugin, for example:

https://github.com/jcupitt/vips-openslideload

**Comment by jcupitt** [comment]
As @remicollet says in #1519, a heic pugin is really necessary due to patent issues. Another reason to work on this feature.

**Comment by remicollet** [comment]
> I don't think this would work on Windows. My understanding is that DLLs do not support backlinking...

Sounds strange, this is exactly how PHP (for ex) works, extensions make call from main core php (or from other extensions)

**Comment by jcupitt** [comment]
I'm probably wrong, but my understanding from talking to a Windows developer friend was that win plugins are usually loaded to the side of the library, as it were. You do it like this:

```
A.exe ----------------> library.dll
    |                       ^
    |                       |
    +-- plugin.dll ---------+
```

So the plugin and the exe link to the library, and the exe loads the plugin.

libvips plugins work like this:

```
executable ------> libvips ------> plugin
                     ^              |
                     +--------------+
```

ie. the plugin needs to call back into the thing that loads it with `dlopen()`. *nix allows backlinking because it has a run-time linker, but win just has the basic dll stub mechanism and it's not possible.

**Comment by cmb69** [comment]
>  *nix allows backlinking because it has a run-time linker, but win just has the basic dll stub mechanism and it's not possible.

At the very least, you could dynamically link, though, i.e. using `GetProcAddress()` to get the addresses of the functions, and then call them.

**Comment by sandstrom** [comment]
As additional context, here are the dependencies on macOS Homebrew for vips.

For us, I think most of these are overkill (for example, I don't think we need imagemagick?). 

We're only using vips to open PNG and JPEG. May use webp in the future, but I don't think we'll ever use TIFF for example. Dynamical loaders would be quite useful!

Thanks for an awesome library! 🏅 Vips is really excellent, we're using it for fast JPEG resize as well as some other image manipulation. 

![image](https://user-images.githubusercontent.com/122287/105494833-a9107580-5cbb-11eb-853c-780f0258ded2.png)

**Comment by jcupitt** [comment]
Yes, the homebrew deps are huge. They have a policy of enabling all optional dependencies so they can serve precompiled binaries as often as possible.

**Comment by kleisauke** [comment]
I started working on this, see PR https://github.com/libvips/libvips/pull/2167. Any testing would be great!
